client/my-sites/plugins/plugins-list/index.jsx

```diff
@@ -0,0 +1,390 @@
+/**
+ * External dependencies
+ */
+import React, { PropTypes } from 'react';
+import classNames from 'classnames';
+import isEmpty from 'lodash/lang/isEmpty';
+
+/**
+ * Internal dependencies
+ */
+import analytics from 'analytics';
+import acceptDialog from 'lib/accept';
+import DisconnectJetpackDialog from 'my-sites/plugins/disconnect-jetpack/disconnect-jetpack-dialog';
+import PluginItem from 'my-sites/plugins/plugin-item/plugin-item';
+import PluginsActions from 'lib/plugins/actions';
+import PluginsListHeader from 'my-sites/plugins/plugin-list-header';
+import PluginsLog from 'lib/plugins/log-store';
+import PluginNotices from 'lib/plugins/notices';
+import SectionHeader from 'components/section-header';
+
+export default React.createClass( {
+	displayName: 'PluginsList',
+
+	mixins: [ PluginNotices ],
+
+	propTypes: {
+		plugins: PropTypes.array,
+		header: PropTypes.string,
+		isWpCom: PropTypes.bool,
+		sites: PropTypes.object,
+		selectedSite: PropTypes.oneOfType( [ PropTypes.bool, PropTypes.object ] ),
+		pluginUpdateCount: PropTypes.number,
+		isPlaceholder: PropTypes.bool,
+	},
+
+	getInitialState() {
+		return {
+			disconnectJetpackDialog: false,
+			bulkManagement: false,
+			selectedPlugins: {}
+		};
+	},
+
+	componentDidMount() {
+		PluginsLog.on( 'change', this.showDisconnectDialog );
+	},
+
+	componentWillUnmount() {
+		PluginsLog.removeListener( 'change', this.showDisconnectDialog );
+	},
+
+	isSelected( plugin ) {
+		return !! this.state.selectedPlugins[ plugin.slug ];
+	},
+
+	togglePlugin( plugin ) {
+		let selectedPlugins = Object.assign( {}, this.state.selectedPlugins );
+		selectedPlugins[ plugin.slug ] = ! this.state.selectedPlugins[ plugin.slug ];
+		this.setState( { selectedPlugins } );
+		analytics.ga.recordEvent( 'Plugins', 'Clicked to ' + this.isSelected( plugin ) ? 'Deselect' : 'Select' + 'Single Plugin', 'Plugin Name', plugin.slug );
+	},
+
+	setBulkSelectionState( plugins, selectionState ) {
+		let slugsToBeUpdated = {};
+		plugins.forEach( plugin => slugsToBeUpdated[ plugin.slug] = selectionState );
+
+		let selectedPlugins = Object.assign( {}, this.state.selectedPlugins, slugsToBeUpdated );
+
+		this.setState( { selectedPlugins } );
```

Comment:
it appears like we are essentially getting a list of slugs and a boolean directive to apply to all of them, then we update the state so that the saved list of plugins now have that boolean. is this correct?

```js
const mergeByKeyWithValue = value => ( obj, key ) => Object.assign( obj, { [ key ]: value } );

const selectedPlugins = plugins
	.map( p => p.slug )
	.reduce( mergeByKeyWithValue( selectionState ), { ...this.state.selectedPlugins } );

this.setState( { selectedPlugins } );
```

In this code I'm actually using the _mutable_ version of `Object.assign()` but since we're mutating the object created from a clone of the existing state it's okay - the mutations never leave the reducer loop. The helper function isn't necessary but I personally like it because it tells me _what_ I'm intending to do. Of course, I tend to use this common vocabulary of a `merge` function when working with reducers. E.g. maybe this approach could have been taken here too to simplify the reducer:

```js
// takes a list of key/value pairs and folds them into an object as key: value entries
const mergeIntoObject = ( obj, [ key, value ] ) => Object.assign( obj, { [key]: value } );

const selectedPlugins = plugins
	.map( p => [ p.slug, selectionState ] )
	.reduce( mergeIntoObject, { ...this.state.selectedPlugins } );

this.setState( { selectedPlugins } );
```

Actually I like this second bit more than my first bit. The input is pushed to the top of the function, the output to the bottom, and the lines in the middle are pretty clear in indicating what they are supposed to accomplish.